[privacy] Add generated privacy keys to GroupDataProvider.

turon opened the pull request:
#### Problem
In order to support future roll-out of privacy, v1.0 nodes need to minimally be able to parse messages that have privacy enabled.

#### Change overview
This PR adds support for deriving the privacy keys associated with a KeySet and adding them to the data store.

#### Testing
Added new coverage for all group credential derivations when an epoch key is added to a KeySet:
- encryption key
- privacy key
- session id (group key hash)

turon commented on src/credentials/GroupDataProviderImpl.cpp:
```diff
@@ -1853,20 +1856,22 @@ CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::DecryptMessage(const ByteSpan
+                                                                  MutableByteSpan & output) const
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    return Crypto::AES_CCM_encrypt(input.data(), input.size(), nullptr, 0, mPrivacyKey, Crypto::kAES_CCM128_Key_Length,
```
```suggestion
    // TODO(#21559): Add AES_CTR_encrypt and use that
    return Crypto::AES_CCM_encrypt(input.data(), input.size(), nullptr, 0, mPrivacyKey, Crypto::kAES_CCM128_Key_Length,
```

turon replied:
```suggestion
    // TODO(#21559): Add AES_CTR_encrypt and use that
    return Crypto::AES_CCM_encrypt(input.data(), input.size(), nullptr, 0, mPrivacyKey, Crypto::kAES_CCM128_Key_Length,
```

turon commented on src/credentials/GroupDataProviderImpl.cpp:
```diff
@@ -1853,20 +1856,22 @@ CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::DecryptMessage(const ByteSpan
+                                                                  MutableByteSpan & output) const
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    return Crypto::AES_CCM_decrypt(input.data(), input.size(), nullptr, 0, nullptr, 0, mPrivacyKey, Crypto::kAES_CCM128_Key_Length,
```
```suggestion
    // TODO(#21559): Add AES_CTR_decrypt and use that
    return Crypto::AES_CCM_decrypt(input.data(), input.size(), nullptr, 0, nullptr, 0, mPrivacyKey, Crypto::kAES_CCM128_Key_Length,
```

tcarmelveilleux commented on src/credentials/GroupDataProviderImpl.cpp:
```diff
@@ -1853,20 +1856,24 @@ CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::DecryptMessage(const ByteSpan
+                                                                  MutableByteSpan & output) const
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    // TODO(#21559): Add AES_CTR_decrypt and use that
```
If this is not used yet, why do we cause anyone to have to pay the RAM cost for this feature?

turon replied:
v1.0 nodes will need to be able to parse v1.0 message format.  That includes messages with privacy, MX, and / or SX enabled.

turon replied:
Each privacy key is re-derived from the encryption_key in `KeySetData::Deserialize` to avoid any growth of `PersistentData`.

turon commented on src/transport/CryptoContext.cpp:
```diff
@@ -253,4 +262,40 @@ CHIP_ERROR CryptoContext::Decrypt(const uint8_t * input, size_t input_length, ui
+
+    const ByteSpan privacytext(input, input_length);
+    MutableByteSpan plaintext(output, input_length);
+    CryptoContext::NonceStorage privacyNonce;
```
Privacy design is based on low probability of nonce overlap (2^-13).  That's why it is call Obfuscation to manage expectations.  Worst case, nonce overlap reveals a small portion of the message header as if privacy is disabled.

rcasallas-silabs commented on src/credentials/GroupDataProvider.h:
```diff
@@ -306,6 +318,9 @@ class GroupDataProvider
     // Key Sets
     //
 
+    static CHIP_ERROR DeriveOperationalKey(const ByteSpan & epoch_key, const ByteSpan & compressed_fabric_id,
```
I can't find any reference of this function. Did you mean `DeriveOperationalCredentials`?

turon replied:
Removed unused method.

rcasallas-silabs commented on src/credentials/GroupDataProviderImpl.h:
```diff
@@ -80,6 +80,8 @@ class GroupDataProviderImpl : public GroupDataProvider
     // Key Sets
     //
 
+    static CHIP_ERROR DeriveOperationalCredentials(const ByteSpan & epoch_key, const ByteSpan & compressed_fabric_id,
```
This function is misplaced, I believe it should go into CHIPCryptoPAL.h

turon replied:
Moved as suggested, which forced moving `GroupOperationalCredentials` structure to `chip::Crypto` as well.

rcasallas-silabs commented on src/credentials/GroupDataProviderImpl.h:
```diff
@@ -169,17 +179,16 @@ class GroupDataProviderImpl : public GroupDataProvider
-                                  const ByteSpan & mic) const override;
-        CHIP_ERROR DecryptPrivacy(MutableByteSpan & header, uint16_t session_id, const ByteSpan & payload,
-                                  const ByteSpan & mic) const override;
+        CHIP_ERROR PrivacyEncrypt(const ByteSpan & input, const ByteSpan & nonce, MutableByteSpan & output) const override;
```
Methods on this interface use the format (Verb)(Noun), if you invert the order, at least you should change the order of the other function to maintain consistency (MessageDecrypt instead of DecryptMessage).

turon replied:
Renamed as advised.

bzbarsky-apple commented on src/credentials/GroupDataProvider.h:
```diff
@@ -28,6 +28,8 @@
 namespace chip {
 namespace Credentials {
 
+using GroupOperationalCredentials = Crypto::GroupOperationalCredentials;
```
Please don't put using declarations not scoped to a specific class in headers.

turon replied:
Removed

bzbarsky-apple commented on src/credentials/GroupDataProviderImpl.cpp:
```diff
@@ -1638,6 +1628,15 @@ CHIP_ERROR GroupDataProviderImpl::GetKeySet(chip::FabricIndex fabric_index, uint
     out_keyset.epoch_keys[2].start_time = keyset.operational_keys[2].start_time;
+
+    // Re-derive privacy keys to minimize storage
+    for (size_t i = 0; i < keyset.keys_count; ++i)
+    {
+        ByteSpan encryption_key(keyset.operational_keys[i].encryption_key, Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES);
+        MutableByteSpan privacy_key(keyset.operational_keys[i].privacy_key, Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES);
+        ReturnErrorOnFailure(Crypto::DeriveGroupPrivacyKey(encryption_key, privacy_key));
```
I don't understand this block.  `keyset` does not escape this function.  It's not used after this block.  Why are we modifying anything about `keyset` here?

turon replied:
Removed and moved to Deserialize.

bzbarsky-apple commented on src/credentials/GroupDataProviderImpl.cpp:
```diff
@@ -1835,38 +1835,43 @@ CHIP_ERROR GroupDataProviderImpl::GetIpkKeySet(FabricIndex fabric_index, KeySet
 
-CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::EncryptPrivacy(MutableByteSpan & header, uint16_t session_id,
-                                                                  const ByteSpan & payload, const ByteSpan & mic) const
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::PrivacyEncrypt(const ByteSpan & input, const ByteSpan & nonce,
```
Can we call this "PrivacyObfuscate" please?

tcarmelveilleux replied:
+1

bzbarsky-apple commented on src/credentials/GroupDataProviderImpl.cpp:
```diff
@@ -1835,38 +1835,43 @@ CHIP_ERROR GroupDataProviderImpl::GetIpkKeySet(FabricIndex fabric_index, KeySet
 
-CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::DecryptPrivacy(MutableByteSpan & header, uint16_t session_id,
-                                                                  const ByteSpan & payload, const ByteSpan & mic) const
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::PrivacyDecrypt(const ByteSpan & input, const ByteSpan & nonce,
```
And PrivacyDeobfuscate?

bzbarsky-apple commented on src/credentials/GroupDataProviderImpl.cpp:
```diff
@@ -1791,11 +1790,12 @@ Crypto::SymmetricKeyContext * GroupDataProviderImpl::GetKeyContext(FabricIndex f
-                                                   key->hash);
+                return mGroupKeyContexPool.CreateObject(
+                    *this, ByteSpan(creds->encryption_key, Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES), creds->hash,
+                    ByteSpan(creds->privacy_key, Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES));
```
I don't see how `privacy_key` can possibly be valid here.  Where are we setting it?  I can't find it.

turon replied:
Should be valid now that privacy key derivation is in Deserialize.

bzbarsky-apple commented on src/credentials/tests/TestGroupDataProvider.cpp:
```diff
@@ -1211,6 +1214,63 @@ static EpochKey kEpochKeys3[] = {
+
+const uint16_t theGroupKeySetTestVectorLength = sizeof(theGroupKeySetTestVector) / sizeof(theGroupKeySetTestVector[0]);
+
+void TestKeySetPrivacy(nlTestSuite * apSuite, void * apContext)
```
This test is not testing anything about the provider behavior.  Shouldn't it test that?

turon replied:
Moved GroupOperationalCredentials tests to crypto.
The key derivation within a KeySet can't be tested directly because the keys aren't accessible beyond the Encrypt/Decrypt API.

rcasallas-silabs commented on src/crypto/CHIPCryptoPAL.h:
```diff
@@ -1523,30 +1538,26 @@ class SymmetricKeyContext
      */
-    virtual CHIP_ERROR EncryptPrivacy(MutableByteSpan & header, uint16_t session_id, const ByteSpan & payload,
-                                      const ByteSpan & mic) const = 0;
+    virtual CHIP_ERROR PrivacyObfuscate(const ByteSpan & input, const ByteSpan & nonce, MutableByteSpan & output) const = 0;
```
What's wrong with "encrypt" / "decrypt"? I know the spec uses the word "obfuscate" in some paragraphs, but I see no reason to use an obscure synonym, specially when we use "encrypt" / "decrypt" everywhere else.

turon replied:
We pivoted to "obfuscate" to provide better clarity on the lack of security guarantees of the privacy header feature, as raised in these two comments:
_Originally posted by @tcarmelveilleux in https://github.com/project-chip/connectedhomeip/pull/21555#discussion_r936967045_
_Originally posted by @bzbarsky-apple in https://github.com/project-chip/connectedhomeip/pull/21555#discussion_r937790096_

Personally, I prefer as you do "PrivacyEncrypt" as that matches the name of the Ch02 crypto primitive in the spec, and the term "privacy" provides enough reference to the feature name to signal associated security caveats.  The preference isn't strong however, and I'm ultimately neutral on which way we go for the naming.

@bzbarsky-apple ?

bzbarsky-apple replied:
I can live with making these operations' names match their spec names.

turon replied:
Ok, renamed to align with Ch02 primitives.

bzbarsky-apple commented on src/credentials/GroupDataProviderImpl.cpp:
```diff
@@ -830,11 +821,14 @@ struct KeySetData : PersistentData<kPersistentBufferMax>
+                VerifyOrReturnError(Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES == encryption_key.size(), CHIP_ERROR_INTERNAL);
+                memcpy(key.encryption_key, encryption_key.data(), encryption_key.size());
+                // Re-derive privacy key from encryption key when loading from storage to save on storage size.
+                MutableByteSpan privacy_key(key.privacy_key, Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES);
```
```suggestion
                MutableByteSpan privacy_key(key.privacy_key);
```
It's already got the right size, and this enforces that we don't get size mismatches between here and the declaration of `privacy_key`.

tcarmelveilleux commented on src/transport/CryptoContext.cpp:
```diff
@@ -147,6 +147,15 @@ CHIP_ERROR CryptoContext::BuildNonce(NonceView nonce, uint8_t securityFlags, uin
+    Encoding::LittleEndian::BufferWriter bbuf(nonce.data(), nonce.size());
+
+    bbuf.Put16(sessionId);
+    bbuf.Put(mac.GetTag(), kPrivacyNonceMicFragmentLen);
```
This does not follow spec!

- The BufferWriter must be Big-Endian (the session ID must be written big-endian)
- The last 11 bytes of the MIC must be used, not the first 11.

See https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/5518

turon replied:
Corrected and added unit test

tcarmelveilleux reviewed (changes requested):
- [ ] Privacy nonce computation is not spec compliant. It also does not have a test.
- [ ] Missing unit tests for privacy obfuscation. This would have caught the above
- [ ] GroupKeyContext::PrivacyEncrypt should be GroupKeyContext::PrivacyObfuscate
- [ ] Need to explain why `AES_CCM_encrypt/AES_CCM_decrypt` is satisfactory to use when spec calls for :
```
Crypto_Privacy_Encrypt(K, M, N) :=
     byte[lengthInBytes(M)]
     AES-CTR-Encrypt(K := K, P := M, N := N)
```
  - AES-CCM does use AES-CTR for the data portion. This must be explained

turon commented:
> * [x]  Privacy nonce computation is not spec compliant. It also does not have a test.
Fixed

> * [ ]  GroupKeyContext::PrivacyEncrypt should be GroupKeyContext::PrivacyObfuscate
We discussed this here, and switched the names to `PrivacyEncrypt` to match Ch03 naming (`Crypto_Privacy_Encrypt`): _Originally posted by @rcasallas-silabs in https://github.com/project-chip/connectedhomeip/pull/21555#discussion_r940391448_

These tasks are left for a follow on PR (issue #21559):
> * [ ]  Need to explain why `AES_CCM_encrypt/AES_CCM_decrypt` is satisfactory to use when spec calls for AES_CTR
(Plan is to add formal AES_CTR crypto primitives.  If calling AES_CCM w/o tags is sufficient, then that will be used.)
> * [ ]  Missing unit tests for privacy obfuscation. This would have caught the above
Yes addition of CTR will come with unit tests.

These tasks are also left as a follow on PR:
* [ ] Add privacy decoding to receive pipeline and packet parsing
* [ ] Add privacy encrypt to transmit pipeline when P flag is enabled (always false for 1.0)
* [ ] Add unit vectors for packet encode / decode with privacy
